File: matlab/codes/plots.m
% Script illustrates how to reproduce figures  
% shown in our paper and on this website

% |xdmd| vs frequency
plot(imag(log(Edmd)),abs(xdmd),'ko')
xlab = xlabel('frequency','interpreter','tex')
set(xlab,'FontName','cmmr10','FontSize',26)
ylab = ylabel('amplitude','interpreter','tex')
set(ylab,'FontName','cmmr10','FontSize',26)
h = get(gcf,'CurrentAxes'); 
set(h,'FontName','cmr10','FontSize',20,'xscale','lin','yscale','lin')

% |xdmd| vs real part
plot(real(log(Edmd)),abs(xdmd),'ko')
xlab = xlabel('real','interpreter','tex')
set(xlab,'FontName','cmmr10','FontSize',26)
ylab = ylabel('amplitude','interpreter','tex')
set(ylab,'FontName','cmmr10','FontSize',26)
h = get(gcf,'CurrentAxes'); 
set(h,'FontName','cmr10','FontSize',20,'xscale','lin','yscale','lin')

% Performance loss for the polished vector of amplitudes vs gamma 
semilogx(answer.gamma,answer.Ploss,'ko','LineWidth',1,'MarkerSize',7)
xlab = xlabel('\gamma','interpreter','tex')
set(xlab,'FontName','cmr10','FontSize',26)
ylab = ylabel('performance loss (%)','interpreter','tex')
set(ylab,'FontName','cmr10','FontSize',26)
h = get(gcf,'CurrentAxes'); 
set(h,'FontName','cmr10','FontSize',20)
axis([answer.gamma(1) answer.gamma(end) 0 1.05*answer.Ploss(end)])

% Number of non-zero amplitudes vs gamma
semilogx(answer.gamma,answer.Nz,'ko','LineWidth',1,'MarkerSize',7)
xlab = xlabel('\gamma','interpreter','tex')
set(xlab,'FontName','cmr10','FontSize',26)
ylab = ylabel('N_z','interpreter','tex')
set(ylab,'FontName','cmr10','FontSize',26)
h = get(gcf,'CurrentAxes'); 
set(h,'FontName','cmr10','FontSize',20)
axis([answer.gamma(1) answer.gamma(end) 0 1.05*answer.Nz(1)])

% Spectrum of DT system for a certain value of gamma
m = 50; 
answer.Nz(m) % number of non-zero amplitudes
ival = find(answer.xsp(:,m));
plot(real(Edmd),imag(Edmd),'ko',real(Edmd(ival)),imag(Edmd(ival)),'r+', ...
    'LineWidth',1,'MarkerSize',7)
xlab = xlabel('Re(\mu_i)','interpreter','tex')
set(xlab,'FontName','cmr10','FontSize',26)
ylab = ylabel('Im(\mu_i)','interpreter','tex')
set(ylab,'FontName','cmr10','FontSize',26)
h = get(gcf,'CurrentAxes'); 
set(h,'FontName','cmr10','FontSize',20,'xscale','lin','yscale','lin')
hold
% plot a unit circle
format compact                    % tighten loose format
format long e                     % make numerical output in double precision
theta = linspace(0,2*pi,100);     % create vector theta
x = cos(theta);                   % generate x-coordinate
y = sin(theta);                   % generate y-coordinate
plot(x,y,'b--','LineWidth',1);    % plot unit circle
axis('equal');
hold

% CT spectrum of channel flow problem for a certain value of gamma
m = 50; 
answer.Nz(m) % number of non-zero amplitudes
ival = find(answer.xsp(:,m));
plot(real(Eos),imag(Eos),'b*', ...
    real(Ect),imag(Ect),'ko', ...
    real(Ect(ival)),imag(Ect(ival)),'r+', ...
    'LineWidth',1.,'MarkerSize',7)
xlab = xlabel('Re(\mu_i)','interpreter','tex')
set(xlab,'FontName','cmr10','FontSize',26)
ylab = ylabel('Im(\mu_i)','interpreter','tex')
set(ylab,'FontName','cmr10','FontSize',26)
h = get(gcf,'CurrentAxes'); 
set(h,'FontName','cmr10','FontSize',20,'xscale','lin','yscale','lin')
axis([-0.8 0.1 -1.1 -0.1])

% |xdmd| and |xpol| vs frequency for a certain value of gamma
% amplitudes in log scale 
m = 50; 
answer.Nz(m) % number of non-zero amplitudes
ival = find(answer.xsp(:,m));
semilogy(imag(log(Edmd)),abs(xdmd),'ko', ...
     imag(log(Edmd(ival))),abs(answer.xpol(ival,m)),'r+', ...
    'LineWidth',1,'MarkerSize',7)
xlab = xlabel('frequency','interpreter','tex')
set(xlab,'FontName','cmr10','FontSize',26)
ylab = ylabel('amplitude','interpreter','tex')
set(ylab,'FontName','cmr10','FontSize',26)
h = get(gcf,'CurrentAxes'); 
set(h,'FontName','cmr10','FontSize',20)

% Determine data for performance loss vs number of dmd modes plots
Nz(1) = answer.Nz(1);
Ploss(1) = answer.Ploss(1);
ind = 1;

for i = 1:length(answer.gamma)-1,
    
    if (answer.Nz(i) == answer.Nz(i+1)),
        
        ind = ind;
        
    else 
        
        ind = ind+1;
        Nz(ind) = answer.Nz(i+1);
        Ploss(ind) = answer.Ploss(i+1);
        
    end
   
    
end

clear ind

% Performance loss vs number of dmd modes
plot(Nz,Ploss,'ko')
xlab = xlabel('number of dmd modes','interpreter','tex')
set(xlab,'FontName','cmr10','FontSize',26)
ylab = ylabel('performance loss (%)','interpreter','tex')
set(ylab,'FontName','cmr10','FontSize',26)
h = get(gcf,'CurrentAxes'); 
set(h,'FontName','cmr10','FontSize',20)
axis([Nz(end) Nz(1) 0 1.05*Ploss(end)])

File: sparse_dmd/plots.py
import functools

import numpy as np
import matplotlib.pyplot as plt


def subplot(plot_function):
    """Wrapper for functions that plot on a matplotlib axes instance
    that autocreates a figure and axes instance if ax is not
    supplied.
    """
    @functools.wraps(plot_function)
    def f(self, ax=None, **kwargs):
        if ax is None:
            fig, ax = plt.subplots()
            return plot_function(self, ax, **kwargs)
        else:
            return plot_function(self, ax, **kwargs)
    return f


class SparsePlots(object):
    def __init__(self, sparsedmd):
        self.dmd = sparsedmd.dmd
        self.sparsedmd = sparsedmd

        self.xdmd = self.dmd.xdmd
        self.Ydmd = self.dmd.Ydmd
        self.Edmd = self.dmd.Edmd
        self.Fdmd = self.dmd.Fdmd

        self.sparse = sparsedmd.sparse

    @subplot
    def xdmd_frequency(self, ax):
        """|xdmd| vs frequency"""
        ax.plot(np.log(self.Edmd).imag, np.abs(self.xdmd), 'ko')
        ax.set_xlabel('frequency')
        ax.set_ylabel('amplitude')

    @subplot
    def xdmd_real(self, ax):
        """|xdmd| vs real part"""
        ax.plot(np.log(self.Edmd).real, np.abs(self.xdmd), 'ko')
        ax.set_xlabel('real')
        ax.set_ylabel('amplitude')

    @subplot
    def performance_loss_gamma(self, ax):
        """Performance loss for the polished vector of amplitudes vs gamma"""
        ax.semilogx(self.sparse.gamma, self.sparse.Ploss, 'ko',
                    linewidth=1, markersize=7)
        ax.set_xlabel(r'$\gamma$')
        ax.set_ylabel('performance loss (%)')
        ax.axis([self.sparse.gamma[0], self.sparse.gamma[-1],
                 0, 1.05 * self.sparse.Ploss[-1]])

    @subplot
    def nonzero_gamma(self, ax):
        """Number of non-zero amplitudes vs gamma"""
        ax.semilogx(self.sparse.gamma, self.sparse.Nz,
                    'ko', linewidth=1, markersize=7)
        ax.set_xlabel(r'$\gamma$')
        ax.set_ylabel(r'$N_z$')
        ax.axis([self.sparse.gamma[0], self.sparse.gamma[-1],
                 0, 1.05 * self.sparse.Nz[0]])

    @subplot
    def spectrum_gamma(self, ax, m=20):
        """Spectrum of DT system for a certain value of gamma"""
        ival = np.where(self.sparse.xsp[:, m])  # non zero amplitudes
        ax.plot(self.Edmd.real, self.Edmd.imag, 'ko',
                self.Edmd[ival].real, self.Edmd[ival].imag, 'r+',
                linewidth=1, markersize=7)
        ax.set_xlabel(r'$Re(\mu_i)$')
        ax.set_ylabel(r'$Im(\mu_i)$')

        # plot a unit circle
        theta = np.linspace(0, 2 * np.pi, 100)  # create vector theta
        x = np.cos(theta)                       # generate x-coordinate
        y = np.sin(theta)                       # generate y-coordinate
        ax.plot(x, y, 'b--', linewidth=1)       # plot unit circle
        ax.axis('equal')

    @subplot
    def xdmd_xpol_frequency(self, ax, m=20):
        """|xdmd| and |xpol| vs frequency for a certain value of
        gamma amplitudes in log scale
        """
        ival = np.where(self.sparse.xsp[:, m])  # non zero amplitudes
        ax.semilogy(np.log(self.Edmd).imag, np.abs(self.xdmd),
                    'ko', linewidth=1, markersize=7)
        ax.semilogy(np.log(self.Edmd[ival]).imag,
                    np.abs(self.sparse.xpol[ival, m].squeeze()),
                    'r+', linewidth=1, markersize=7)
        ax.set_xlabel('frequency')
        ax.set_ylabel('amplitude')

    @subplot
    def performance_loss_nmodes(self, ax):
        """Performance loss vs number of dmd modes
        """
        # we only care about the performance loss
        # when the number of modes changes
        n_modes = self.sparse.Nz
        n_change = np.where(np.diff(n_modes))

        Nz = self.sparse.Nz[n_change]
        Ploss = self.sparse.Ploss[n_change]

        ax.plot(Nz, Ploss, 'ko')
        ax.set_xlabel('number of dmd modes')
        ax.set_ylabel('performance loss (%)')
        ax.axis([Nz[-1], Nz[0], 0, 1.05 * Ploss[-1]])
